test/lib/query_parser_test.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+require 'query_parser'
+
+class QueryParserTest < ActiveSupport::TestCase
+  setup do
+    @parser = Insights::API::Common::QueryParser.new(
+      Rack::Utils.default_query_parser
+    )
+  end
+
+  test '#parse_nested_query' do
+    {
+      'foo=bar' => { foo: 'bar' },
```

Comment:
How about one that encodes the slash `/`? `tags=insights-client%2Fenv%3Dprod`

Reply:
Sure, I can cover that as well.

Reply:
I took a look at the docs, which I assume is mostly under [this header](https://platform-docs.cloud.paas.psi.redhat.com/backend/inventory.html?highlight=cyndi#string-representation). It does not say that the delimiters must be escaped, it simply says that `=` or `/` in the namespace, key, or value must be escaped. I haven't found docs that say what you're claiming:

>  the Insights API accepts encoded / and = if they are part of a namespace, key or value, **but not when delimiting these**

The docs state:

> Note that in transit (e.g. when sent as an HTTP query parameter value) the tag will be URI-encoded again (as required by RFC3986). This happens automatically in modern HTTP browsers and http client libraries. This should not be confused with the aforementioned encoding of the tag segments described above which the client must always do.

Again, I found this discrepancy by comparing how the Inventory API works, and it accepts encoded delimiters.

Reply:
Okay, so how should we handle if the delimiters and the `/` and `=` symbols in the fields are both escaped? There's no way to distinguish them, only if we allow double-escaping, but should we also expect double escaping if the delimiters are in the unescaped form? 

Btw I've been trying to follow [this one](https://clouddot.pages.redhat.com/docs/dev/services/inventory.html#_string_representation)

Reply:
Here is an example request from the Inventory UI:

```
https://cloud.redhat.com/api/inventory/v1/hosts?per_page=50&page=1&order_by=updated&order_how=DESC&staleness=fresh&staleness=stale&tags=satellite%2Flifecycle_environment%3DLibrary&registered_with=insights&fields%5Bsystem_profile%5D%5B%5D=operating_system
```

I'm not sure I understand the issue with parsing the tags. The first `/` you get, encoded or not, will be the delimiter between the namespace and key. The first `=` you get after that, encoded or not, will be the delimiter between the key and the value. Any further `=` you get will be part of the value, and btw must be encoded.

Reply:
> The first / you get, encoded or not, will be the delimiter between the namespace and key.

So there's no way for the namespace to contain a `/`? :thinking: 

 > The first = you get after that, encoded or not, will be the delimiter between the key and the value

So there's no way for the key to contain a `=`?

>  Any further = you get will be part of the value, and btw must be encoded.

So the key can contain a `/` symbol and the value can contain both `/` and `=`?

This is confusing :thinking: :rofl:

Reply:
Please check with the Inventory API to verify your questions.

> So there's no way for the namespace to contain a `/`?

```
$ http https://cloud.redhat.com/api/inventory/v1/hosts tags==satellite/foo/lifecycle_environment=Library # NB: httpie escapes this query param
HTTP/1.1 400 Bad Request

{
    "detail": "'satellite/foo/lifecycle_environment=Library' does not match '^([^=/]+/)?[^=/]+(=[^=/]+)?$'\n\nFailed validating 'pattern' in schema['items']:\n    {'pattern': '^([^=/]+/)?[^=/]+(=[^=/]+)?$', 'type': 'string'}\n\nOn instance[0]:\n    'satellite/foo/lifecycle_environment=Library'",
    "status": 400,
    "title": "Bad Request",
    "type": "about:blank"
}
```

Nope, seems not. How nice of them to provide the regex :laughing: They should have included it in the docs!

> So there's no way for the key to contain a `=`

Nope, seems not. Nor can it contain a `/`

> So the key can contain a / symbol and the value can contain both `/` and `=`?

The key cannot contain a `/` according to the regex and testing with their API. The value can contain `=` but not `/` according to their API and the regex.

All that said, if you double escape anything, it's fair game.

i.e. `?tags=satellite%2Flifecycle_environment%3DLi%%3Dbrary`

Reply:
@akofink from [this page](https://clouddot.pages.redhat.com/docs/dev/services/inventory.html#_string_representation):
> The fact that a namespace, key or value can contain / and = characters creates a potential ambiguity when parsing a tag in the string representation.

I am just more confused now :disappointed:

Reply:
They have to be escaped. That's the very next sentence:

> In order to avoid this ambiguity it is required that / and = characters that occur in the namespace, key or value are escaped using RFC3986 (URI encoding).

This means the API requester (aka the client) must escape these values if they're in the namespace, key, or value. I demonstrated that above with the Inventory API. I definitely agree with you that it's confusing :laughing:

Reply:
Okay, now I just understand it less than before :sweat:

Reply:
It seems that we are confusing client side escaping with server side. Clients should escape it one their side, escaping the query string's param. It should be client's responsibility to do so.

If we are talking about using these two characters as a value within the tags, then the resulting query string sent by a client is double encoded. Server should get it in as if the URL wasn't decoded, but just for the value part.

```
// client sends
console.log(encodeURIComponent(encodeURIComponent('/=')));
> "%252F%253D"
// server gets 
console.log(encodeURIComponent('/='));
> "%2F%3D"
```

@akofink this case seems to be validate a few lines below.

Reply:
Like I keep saying, please validate this works the same as inventory's tags param:

```
GET inventory/v1/hosts?tags=satellite/lifecycle_environment=Library # works
GET compliance/v1/systems?tags=satellite/lifecycle_environment=Library # works

GET inventory/v1/hosts?tags=satellite%2Flifecycle_environment%3DLibrary # works
GET compliance/v1/systems?tags=satellite%2Flifecycle_environment%3DLibrary # doesn't work
```

The fact that the tests pass indicate this case is not actually covered.

Reply:
These four should be equivalent:
```
tags=satellite/lifecycle_environment=Library
tags=satellite%2Flifecycle_environment=Library
tags=satellite/lifecycle_environment%3DLibrary
tags=satellite%2Flifecycle_environment%3DLibrary
```

What we are missing is a decode of a raw query string.
```
> params
=> {"tags"=>"satellite/lifecycle_environment=Library"}
> query_string
=> "tags=satellite%2Flifecycle_environment%3DLibrary"
> Rack::Utils.unescape(query_string)
=> "tags=satellite/lifecycle_environment=Library
```

Speaking with @skateman, we kinda agree that replacing `tags=` to `tags[]=` (or allowing array params w/o `[]`)  could solve the first parsing problem. Then the values can be handed in the controller (or middleware).